rcl/src/rcl/security.c

```diff
@@ -0,0 +1,350 @@
+  if (NULL != get_env_error_str) {
+    RCL_SET_ERROR_MSG_WITH_FORMAT_STRING(
+      "Error getting env var '" RCUTILS_STRINGIFY(ROS_SECURITY_STRATEGY_VAR_NAME) "': %s\n",
+      get_env_error_str);
+    return RCL_RET_ERROR;
+  }
+
+  *policy = (0 == strcmp(ros_enforce_security, "Enforce")) ?
+    RMW_SECURITY_ENFORCEMENT_ENFORCE : RMW_SECURITY_ENFORCEMENT_PERMISSIVE;
+  return RCL_RET_OK;
+}
+
+/**
+ * A security lookup function takes in the node/context's name, namespace, a security root directory and an allocator;
+ *  It returns the relevant information required to load the security credentials,
+ *  which is currently a path to a directory on the filesystem containing DDS Security permission files.
+ */
+typedef char * (* security_lookup_fn_t) (
+  const char * name,
+  const char * namespace_,
+  const char * ros_secure_root_env,
+  const rcl_allocator_t * allocator
+);
+
+char * exact_match_lookup(
+  const char * name,
+  const char * namespace_,
+  const char * ros_secure_root_env,
+  const rcl_allocator_t * allocator
+);
+
+char * prefix_match_lookup(
+  const char * name,
+  const char * namespace_,
+  const char * ros_secure_root_env,
+  const rcl_allocator_t * allocator
+);
+
+const security_lookup_fn_t g_security_lookup_fns[] = {
+  NULL,
+  exact_match_lookup,
+  prefix_match_lookup,
+};
+
+typedef enum ros_security_lookup_type_e
+{
+  ROS_SECURITY_LOOKUP_OVERRIDE = 0,
+  ROS_SECURITY_LOOKUP_MATCH_EXACT = 1,
+  ROS_SECURITY_LOOKUP_MATCH_PREFIX = 2,
+} ros_security_lookup_type_t;
+
+const char * const g_security_lookup_type_strings[] = {
+  "DIRECTORY_OVERRIDE",
+  "MATCH_EXACT",
+  "MATCH_PREFIX"
+};
+
+/// Return the directory that most closely matches a given name (longest-prefix match),
+/// scanning under base_dir.
+/**
+ * By using a prefix match, a name "my_name_123" will be able to load and use the
+ * directory "my_name" if no better match exists.
+ * \param[in] base_dir
+ * \param[in] name
+ * \param[out] matched_name must be a valid memory address allocated with at least
+ * _TINYDIR_FILENAME_MAX characters.
+ * \return true if a match was found
+ */
+static bool get_best_matching_directory(
+  const char * base_dir,
+  const char * name,
+  char * matched_name)
+{
+  size_t max_match_length = 0;
+  tinydir_dir dir;
+  if (NULL == base_dir || NULL == name || NULL == matched_name) {
+    return false;
+  }
+  if (-1 == tinydir_open(&dir, base_dir)) {
+    return false;
+  }
+  while (dir.has_next) {
+    tinydir_file file;
+    if (-1 == tinydir_readfile(&dir, &file)) {
+      goto cleanup;
+    }
+    if (file.is_dir) {
+      size_t matched_name_length = strnlen(file.name, sizeof(file.name) - 1);
+      if (0 ==
+        strncmp(
+          file.name, name, matched_name_length) &&
+        matched_name_length > max_match_length)
+      {
+        max_match_length = matched_name_length;
+        memcpy(matched_name, file.name, max_match_length);
```

Comment:
@ivanpauno same as in other PRs, use `strncpy` or `rcutils_format_string` or `rcutils_strdup` instead of plain old `memcpy`

Reply:
I didn't modify, I just moved it.
In this case, `rcutils_strdup` doesn't apply.
I will replace `memcpy` with `strncpy`, which is more appropiate here.

Reply:
See https://github.com/ros2/rcl/pull/515/commits/a905ae9dc01d24d4b09f0999f538b61a0e83315e.